### ISSUE TITLE
How to handle subcarrier indices？

### ISSUE THREAD

**Issue by wangjian-hello** [JJ8428/ESP32_CSI_NLOS_Classifier#2]
Hello, I have another question regarding the subcarrier index of LLTF, HT-LTF, and SUBC-HT-LTF in the raw CSI data. I noticed that the code mentions it is adapted from:
https://github.com/espressif/esp-csi/blob/master/examples/get-started/tools/csi_data_read_parse.py. I carefully read through this code, but I personally believe this part of the processing is incorrect. According to the official ESP32 documentation [here](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/wifi.html#wi-fi-channel-state-information), in the case of below, HT, 40 MHz, non-STBC, as per the table in the documentation, the CSI for LLTF corresponds to subcarrier indices 0 to 63 in the raw CSI data (indices 0–63), which also correspond to the subcarrier indices 0–63. There are no negative subcarrier indices. For HT-LTF, its CSI starts after index 63, i.e., after the CSI corresponding to LLTF. The first part corresponds to the positive subcarrier indices, and the second part corresponds to the negative subcarrier indices. When parsing, these two parts should be reversed and then concatenated to obtain the correct order of HT-LTF CSI. There is no CSI corresponding to SUBC-HT-LTF. There are a total of 384 bytes, with each CSI occupying 2 bytes, so there are 192 CSIs in total. LLTF occupies the first 64 CSIs, and HT-LTF occupies the next 128 CSIs. The CSI corresponding to HT-LTF should be obtained by reversing and concatenating the two parts. For the valid subcarrier indices of each part, they should be processed according to the table below.

**Comment by wangjian-hello** [comment]
![image](https://github.com/user-attachments/assets/4bc50435-a916-468b-8cca-b4bd2cd92d9d)
![image](https://github.com/user-attachments/assets/fc6355b3-67dc-419f-acc0-5a87e86686a8)

**Comment by JJ8428** [comment]
Yes, I used the same documentation, and I blindly followed the code you linked above and trusted it was correct given that it is Espressif's github repo. I think you should make an issue on their repo and see what they say.

If I may ask, where do you see the reversing step? I understand the concatenation, but not the reversing.

**Comment by JJ8428** [comment]
Btw, the code for that link you provided used to say this. I deliberately remember copy and pasting this directly from the example code.

'''
Adapted from:
https://github.com/espressif/esp-csi/blob/master/examples/get-started/tools/csi_data_read_parse.py
'''
# LLTF
lltf_subcarrier_index = []
lltf_subcarrier_index += [i for i in range(6, 32)]
lltf_subcarrier_index += [i for i in range(33, 59)]
lltf_subcarrier_len = len(lltf_subcarrier_index)

# HT-LTF
htltf_subcarrier_index = []
htltf_subcarrier_index += [i for i in range(66, 94)]
htltf_subcarrier_index += [i for i in range(95, 123)]
htltf_subcarrier_len = len(htltf_subcarrier_index)

# STBC-HT-LTF
stbchtltf_subcarrier_index = []
stbchtltf_subcarrier_index += [i for i in range(134, 162)]
stbchtltf_subcarrier_index += [i for i in range(163, 191)]
stbchtltf_subcarrier_len = len(stbchtltf_subcarrier_index)

Like you pointed out, its possible someone fixed that error after I moved from this project.

**Comment by wangjian-hello** [comment]
The subcarrier indices for HT-LTF in the table are from 0 to 63, and from -64 to -1. The positive indices are written at the front, and the negative indices are written at the back. If no reversal is needed, I believe they should be written as -64 to 63, rather than being split into two parts.

**Comment by wangjian-hello** [comment]
> Btw, the code for that link you provided used to say this. I deliberately remember copy and pasting this directly from the example code.顺便说一句，你提供的那个链接的代码曾经这样说。我特意记住直接从示例代码中复制和粘贴它。
> 
> ''' Adapted from:  改编自： https://github.com/espressif/esp-csi/blob/master/examples/get-started/tools/csi_data_read_parse.py '''
> 
> # LLTF
> lltf_subcarrier_index = [] lltf_subcarrier_index += [i for i in range(6, 32)]lltf_subcarrier_index += [i for i in range（6，32）] lltf_subcarrier_index += [i for i in range(33, 59)]lltf_subcarrier_index += [i for i in range（33，59）] lltf_subcarrier_len = len(lltf_subcarrier_index)lltf_subcarrier_len（lltf_subcarrier_index）
> 
> # HT-LTF  HT—LTF
> htltf_subcarrier_index = [] htltf_subcarrier_index += [i for i in range(66, 94)]htltf_subcarrier_index += [i for i in range（66，94）] htltf_subcarrier_index += [i for i in range(95, 123)]htltf_subcarrier_index += [i for i in range（95，123）] htltf_subcarrier_len = len(htltf_subcarrier_index)htltf_subcarrier_len = len（htltf_subcarrier_index）
> 
> # STBC-HT-LTF
> stbchtltf_subcarrier_index = [] stbchtltf_subcarrier_index += [i for i in range(134, 162)]stbchtltf_subcarrier_index +=[i for i in range（134，162）] stbchtltf_subcarrier_index += [i for i in range(163, 191)]stbchtltf_subcarrier_index +=[i for i in range（163，191）] stbchtltf_subcarrier_len = len(stbchtltf_subcarrier_index)stbchtltf_subcarrier_len = len（stbchtltf_subcarrier_index）
> 
> Like you pointed out, its possible someone fixed that error after I moved from this project.就像你指出的那样，在我离开这个项目后，可能有人修复了这个错误。

![image](https://github.com/user-attachments/assets/008429c1-e48b-4959-a58f-cab9574cbd78)
Unfortunately, the code in the official repository still handles the indices in this way.

**Comment by JJ8428** [comment]
Yeah I clicked on the link and saw that STBCH indices longer exist. This bums me out since this implies a lot of my analysis was possibly on completely misinterpreted data. 

Just a follow up questions then from what you see, is STBCHTLTF not picked up by the board then? I was thinking about picking this project up to use Wifi CSI to perform human pose/gesture detection? It seems that I will have to do it with only HTLTF and LLTF indices only.

Btw, I really suggest you make an issue on the Espressif repo. If this is correct, this could mean a lot to many people.
